### PR TITLE
Get rid of dependencies on Qt private headers

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: qeventdispatcher_epoll.cpp [hunk 1]
```diff
@@ -384,16 +431,15 @@ bool QEventDispatcherEpoll::processEvents(QEventLoop::ProcessEventsFlags flags)
     d->interrupt = false;
 
     // we are awake, broadcast it
-    emit awake();
-    QCoreApplicationPrivate::sendPostedEvents(0, 0, d->threadData);
+    Q_EMIT awake();
```
Comment (ghost) [inline qeventdispatcher_epoll.cpp:434]: `Q_EMIT` instead of `emit` for code that uses no_keywords setting

File: qeventdispatcher_epoll.cpp [hunk 2]
```diff
@@ -384,16 +431,15 @@ bool QEventDispatcherEpoll::processEvents(QEventLoop::ProcessEventsFlags flags)
...
 
     int nevents = 0;
-    const bool canWait = (d->threadData->canWait
-                          && !d->interrupt
+    const bool canWait = (!d->interrupt
                           && (flags & QEventLoop::WaitForMoreEvents));
 
     if (canWait)
-        emit aboutToBlock();
+        Q_EMIT aboutToBlock();
```
Comment (ghost) [inline qeventdispatcher_epoll.cpp:442]: `Q_EMIT` instead of `emit` for code that uses no_keywords setting

File: qeventdispatcher_epoll.h [hunk 3]
```diff
@@ -11,10 +11,9 @@
 
 #include <QtCore/qabstracteventdispatcher.h>
 #include <QtCore/qlist.h>
-#include <private/qabstracteventdispatcher_p.h>
-#include <private/qcore_unix_p.h>
-#include <private/qpodlist_p.h>
```
Comment (ghost) [inline qeventdispatcher_epoll.h:16]: No symbols were referenced from this header

File: qeventdispatcher_epoll.h [hunk 4]
```diff
@@ -100,12 +99,12 @@ class Q_CORE_EXPORT QEventDispatcherEpoll : public QAbstractEventDispatcher
 
 };
 
-class Q_CORE_EXPORT QEventDispatcherEpollPrivate : public QAbstractEventDispatcherPrivate
+class Q_CORE_EXPORT QEventDispatcherEpollPrivate
```
Comment (ghost) [inline qeventdispatcher_epoll.h:102]: No need to inherit QAbstractEventDispatcherPrivate as none of its data/functions are used

File: qeventdispatcher_epoll.h [hunk 5]
```diff
@@ -113,8 +112,6 @@ class Q_CORE_EXPORT QEventDispatcherEpollPrivate : public QAbstractEventDispatch
     bool mainThread;
     int thread_pipe[2];
 
-    // highest fd for all socket notifiers
-    int sn_highest;
```
Comment (ghost) [inline qeventdispatcher_epoll.h:117]: `sn_highest` makes sense only for `select()`, not for `epoll()`

File: qeventdispatcher_epoll.cpp [hunk 6]
```diff
@@ -26,13 +23,71 @@
 
 #include <QDebug>
```
Comment (ghost) [inline qeventdispatcher_epoll.cpp:25]: The block below comes from private/qcore_unix_p.h

**Comment by lobbin** [comment]
Looks fine, just need to discuss this with my colleagues on monday :)

**Comment by lobbin** [comment]
Finally we had time to discuss this. The problem as we see it is the block copied from qcore_unix_p.h. What happens if these functions are changed? Would it be possible to make this a compile time option?

**Comment by ghost** [comment]
`qt_safe_read()` and `qt_safe_write()` will not change — all they do is restart `read()` or `write()` in case they were interrupted with a signal (`errno == EINTR`). Probably every *nix developer has to write similar wrappers around `read()` and `write()`.

`qt_safe_pipe()` creates a communication object which is used in `wakeup()` function to wake up the dispatcher. All this function does is creates a pipe using `pipe2()` if it is available or `pipe()` if it is not and makes both ends of the pipe non-blocking and close-on-exec.

Since the custom event dispatcher in Qt4 has to be instantiated before QCoreApplication, the code can be simplified to

``` c++
static inline int qt_safe_pipe(int pipefd[2], int flags = 0)
{
    int ret = ::pipe(pipefd);
    if (ret == -1)
        return -1;

    ::fcntl(pipefd[0], F_SETFD, FD_CLOEXEC);
    ::fcntl(pipefd[1], F_SETFD, FD_CLOEXEC);

    if (flags & O_NONBLOCK) {
        ::fcntl(pipefd[0], F_SETFL, ::fcntl(pipefd[0], F_GETFL) | O_NONBLOCK);
        ::fcntl(pipefd[1], F_SETFL, ::fcntl(pipefd[1], F_GETFL) | O_NONBLOCK);
    }

    return 0;
}
```

and the block from `#if defined(Q_OS_LINUX) && defined(O_CLOEXEC)` till `#endif` can be dropped.

The reason why I wanted to get rid of the private headers is that if your app wants to use them, you have to add this to the project file:

```
QT += core-private
```

which then generates this warning:

```
Project MESSAGE: This project is using private headers and will therefore be tied to this specific Qt module build version.
Project MESSAGE: Running this project against other versions of the Qt modules may crash at any arbitrary point.
Project MESSAGE: This is not a bug, but a result of using Qt internals. You have been warned!
```

Thus, even if Qt's ABI does not change on upgrade (e.g., 4.8.1 => 4.8.2) you will have to recompile the entire application.

But if you want, I can prepare a patch that uses `#ifdef`'s to control whether private headers should be used.

Please let me know.

Thank you,
Vladimir

**Comment by lobbin** [comment]
If you could prepare a patch that uses #ifdef's, that would be super :+1:

**Comment by ghost** [comment]
OK, I will close this request then and prepare four patches:
1. emit => Q_EMIT (to allow `CONFIG += no_keywords`)
2. Remove select()-specifc code
3. Remove private headers which can be safely removed
4. Add #ifdefs for the other privates.

**Pull request closed without merging**